src/common/utxobased/engine/makeUtxoEngineState.ts

```diff
@@ -736,55 +731,42 @@ export const pickNextTask = async (
   }
 
   // Check if there are any addresses pending to be subscribed
-  if (
-    Object.keys(addressSubscribeCache).length > 0 &&
-    !taskCache.addressWatching
-  ) {
+  if (Object.keys(addressSubscribeCache).length > 0) {
+    // These are addresse to which the server has not subscribed
+    const newAddress: string[] = []
     const blockHeight = serverStates.getBlockHeight(uri)
-    // Loop each address that needs to be subscribed
+
+    // Loop each address in the cache
     for (const [address, state] of Object.entries(addressSubscribeCache)) {
       // Add address in the cache to the set of addresses to watch
-      const { path, processing: subscribed } = state
-      // only process newly watched addresses
-      if (subscribed) continue
-      if (path != null) {
-        // Add the newly watched addresses to the UTXO cache
-        addressUtxoCache[address] = {
-          processing: false,
-          path
-        }
-        await addToAddressTransactionCache(
-          args,
-          address,
-          path.format,
-          path.changeIndex,
-          blockHeight,
-          addressTransactionCache
-        )
+      if (!serverStates.serverIsAwareOfAddress(uri, address)) {
+        newAddress.push(address)
       }
+
+      // Add to the addressTransactionCache if they're not yet added:
+      // Only process newly watched addresses
+      if (state.processing) continue
+      // Add the newly watched addresses to the UTXO cache
+      addressUtxoCache[address] = {
+        processing: false,
+        path: state.path
+      }
+      await addToAddressTransactionCache(
+        args,
+        address,
+        state.path.format,
+        state.path.changeIndex,
+        blockHeight,
+        addressTransactionCache
+      )
       state.processing = true
     }
 
-    taskCache.addressWatching = true
-
-    const queryTime = Date.now()
-    const deferred = new Deferred<unknown>()
-    deferred.promise
-      .then(() => {
-        pluginState.serverScoreUp(uri, Date.now() - queryTime)
-      })
-      .catch(() => {
-        taskCache.addressWatching = false
-      })
-    deferred.promise.catch(() => {
-      taskCache.addressWatching = false
-    })
-    serverStates.watchAddresses(
-      uri,
-      Array.from(Object.keys(addressSubscribeCache)),
-      deferred
-    )
-    return true
+    // Subscribe to any new addresses
+    if (newAddress.length > 0) {
+      serverStates.watchAddresses(uri, newAddress)
+      return true
```

Comment:
Is there a different meaning if the pickNextTask returns an undefined or true?
If so, you might want to move the true after the if.

But again, I don't fully understand how the return value from pickNextTask is used.

Reply:
`true` means to continue to pick next task immediately after finishing; `false` or `undefined` exits from the `doWakeUp` loop (I'm assuming this is a sort of going to sleep mechanism).

The reason why we need to return in the `if` is because the outer block is not a condition which will ever be false, so it will halt the entire pick next task loop.